constraint/pkg/client/client.go

```diff
@@ -359,418 +187,404 @@ func (c *Client) AddTemplate(templ *templates.ConstraintTemplate) (*types.Respon
-	return constraintPathMerge(target, createConstraintGKSubPath(gk))
-}
-
-// constraintPathMerge is a shared function for creating constraint paths to
-// ensure uniformity, it is not meant to be called directly.
-func constraintPathMerge(target, subpath string) string {
-	return "/" + path.Join("constraints", target, subpath)
+	return template.getTemplate(), nil
 }
 
 // getTemplateEntry returns the template entry for a given constraint.
-func (c *Client) getTemplateEntry(constraint *unstructured.Unstructured, lock bool) (*templateEntry, error) {
-	kind := constraint.GetKind()
-	if kind == "" {
-		return nil, fmt.Errorf("%w: kind missing from Constraint %q",
-			crds.ErrInvalidConstraint, constraint.GetName())
-	}
+func (c *Client) getTemplateForKind(kind string) (*templateClient, error) {
+	name := strings.ToLower(kind)
 
-	group := constraint.GroupVersionKind().Group
-	if group != constraints.Group {
-		return nil, fmt.Errorf("%w: wrong API Group for Constraint %q, got %q but need %q",
-			crds.ErrInvalidConstraint, constraint.GetName(), group, constraints.Group)
-	}
+	template := c.templates[name]
 
-	if lock {
-		c.mtx.RLock()
-		defer c.mtx.RUnlock()
+	if template == nil {
+		return nil, templateNotFound(name)
 	}
 
-	entry, ok := c.templates[templateKeyFromConstraint(constraint)]
-	if !ok {
-		var known []string
-		for k := range c.templates {
-			known = append(known, string(k))
-		}
-
-		return nil, fmt.Errorf("%w: Constraint kind %q is not recognized, known kinds %v",
-			ErrMissingConstraintTemplate, kind, known)
-	}
-
-	return entry, nil
+	return template, nil
 }
 
 // AddConstraint validates the constraint and, if valid, inserts it into OPA.
 // On error, the responses return value will still be populated so that
 // partial results can be analyzed.
 func (c *Client) AddConstraint(ctx context.Context, constraint *unstructured.Unstructured) (*types.Responses, error) {
+	resp := types.NewResponses()
+
 	c.mtx.Lock()
 	defer c.mtx.Unlock()
 
-	resp := types.NewResponses()
-	entry, err := c.getTemplateEntry(constraint, false)
+	err := c.validateConstraint(constraint)
 	if err != nil {
 		return resp, err
 	}
 
-	subPath, err := createConstraintSubPath(constraint)
+	template, err := c.getTemplateForKind(constraint.GetKind())
 	if err != nil {
-		return resp, fmt.Errorf("creating Constraint subpath: %w", err)
-	}
-
-	// return immediately if no change
-	cached, err := c.getConstraintNoLock(constraint)
-	if err == nil && constraintlib.SemanticEqual(cached, constraint) {
-		for _, target := range entry.Targets {
-			resp.Handled[target] = true
-		}
-		return resp, nil
+		return resp, err
 	}
 
-	if err := c.validateConstraint(constraint, false); err != nil {
+	changed, err := template.AddConstraint(constraint)
+	if err != nil {
 		return resp, err
 	}
-	if err := c.driver.AddConstraint(ctx, constraint); err != nil {
-		return resp, err
+
+	if changed {
+		err = c.driver.AddConstraint(ctx, constraint)
+		if err != nil {
```

Comment:
Either:

1. We need to restart Policy Controller because the Rego inmem storage has gotten into a state where it no longer allows successful writes.
2. The context was cancelled (e.g. timed out)

Reply:
Also - it's okay for the state to be kept consistent (Constraint in both Driver and per-Template cache) if removal from the driver fails due to context cancellation.

Reply:
IIRC at one point we talked about panic-ing if modifying the driver's cache fails? Mainly b/c such failures are unexpected (need to restart PC) and in the case of multi-target it makes it hard/impossible to have consistent state.

Reply:
Made Client no longer allow concurrent updates. While we talked about panicking if modifying the driver's cache fails, it can fail for reasons either meaning Frameworks should be restarted or due to a simple context cancellation/timeout. These modifications aren't necessary for this PR and can be done later. This PR doesn't change how that situation is handled from the existing behavior of just reporting when this happens.

Reply:
I'd rather leave behavior as-is in regards to failing to update the OPA storage, rather than potentially introduce a case where Gatekeeper constantly restarts due to context cancellations being much more common than we anticipate for certain users.